File: legacy/src/main/rs/emulate/legacy/config/item/ItemStack.java
package rs.emulate.legacy.config.item;

import rs.emulate.shared.util.DataBuffer;
import rs.emulate.util.Assertions;

/**
 * A utility class containing an item stack amount and its corresponding model id.
 *
 * @author Major
 */
public final class ItemStack {

	/**
	 * The empty ItemStack, used as the default value.
	 */
	public static final ItemStack EMPTY = new ItemStack();

	/**
	 * Decodes an ItemStack from the specified {@link DataBuffer}.
	 *
	 * @param buffer The Buffer.
	 * @return The ItemStack.
	 */
	public static ItemStack decode(DataBuffer buffer) {
		int amount = buffer.getUnsignedShort();
		int model = buffer.getUnsignedShort();

		return new ItemStack(amount, model);
	}

	/**
	 * Encodes the specified ItemStack into the specified {@link DataBuffer}.
	 *
	 * @param buffer The Buffer.
	 * @param stack The ItemStack.
	 */
	public static void encode(DataBuffer buffer, ItemStack stack) {
		buffer.putShort(stack.getAmount()).putShort(stack.getModel());
	}

	/**
	 * The stack amount.
	 */
	private final int amount;

	/**
	 * The model id.
	 */
	private final int model;

	/**
	 * Creates the ItemStack.
	 *
	 * @param amount The stack amount.
	 * @param model The model id.
	 */
	public ItemStack(int amount, int model) {
		Assertions.checkPositive(amount, "Amount must be greater than 0, received " + amount + ".");
		Assertions.checkNonNegative(model, "Model id must be greater than or equal to 0, received " + model + ".");
		this.amount = amount;
		this.model = model;
	}

	/**
	 * Creates an ItemStack with default values.
	 */
	private ItemStack() {
		amount = 0;
		model = -1;
	}

	/**
	 * Gets the amount of this ItemStack.
	 *
	 * @return The amount.
	 */
	public int getAmount() {
		return amount;
	}

	/**
	 * Gets the model of this ItemStack.
	 *
	 * @return The model.
	 */
	public int getModel() {
		return model;
	}

}

File: legacy/src/main/kotlin/rs/emulate/legacy/config/item/ItemStack.kt
package rs.emulate.legacy.config.item

import rs.emulate.shared.util.DataBuffer

/**
 * A utility class containing an item stack amount and its corresponding model id.
 */
class ItemStack {

    /**
     * The stack amount.
     */
    val amount: Int

    /**
     * The model id.
     */
    val model: Int

    /**
     * Creates the ItemStack.
     */
    constructor(amount: Int, model: Int) {
        require(amount.toLong() > 0) { "Amount must be greater than 0, received $amount." }
        require(model.toLong() >= 0) { "Model id must be greater than or equal to 0, received $model." }
        this.amount = amount
        this.model = model
    }

    /**
     * Creates an ItemStack with default values.
     */
    private constructor() {
        amount = 0
        model = -1
    }

    companion object {

        /**
         * The empty ItemStack, used as the default value.
         */
        val EMPTY = ItemStack()

        /**
         * Decodes an ItemStack from the specified [DataBuffer].
         */
        fun decode(buffer: DataBuffer): ItemStack {
            val amount = buffer.getUnsignedShort()
            val model = buffer.getUnsignedShort()

            return ItemStack(amount, model)
        }

        /**
         * Encodes the specified ItemStack into the specified [DataBuffer].
         */
        fun encode(buffer: DataBuffer, stack: ItemStack): DataBuffer {
            return buffer.putShort(stack.amount)
                .putShort(stack.model)
        }
    }

}
